runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -364,22 +292,13 @@ void outputWindowedValue(WindowedValue<OutputT> windowedElem) {
         Instant timestamp,
         Collection<? extends BoundedWindow> windows,
         PaneInfo pane) {
-      outputWindowedValue(tag, makeWindowedValue(output, timestamp, windows, pane));
+      checkNotNull(windows);
+      checkNotNull(timestamp);
+      outputWindowedValue(tag, WindowedValue.of(output, timestamp, windows, pane));
     }
 
     private <T> void outputWindowedValue(TupleTag<T> tag, WindowedValue<T> windowedElem) {
-      if (!outputTags.contains(tag)) {
-        // This tag wasn't declared nor was it seen before during this execution.
-        // Thus, this must be a new, undeclared and unconsumed output.
-        // To prevent likely user errors, enforce the limit on the number of side
-        // outputs.
-        if (outputTags.size() >= MAX_SIDE_OUTPUTS) {
-          throw new IllegalArgumentException(
-              "the number of outputs has exceeded a limit of " + MAX_SIDE_OUTPUTS);
-        }
-        outputTags.add(tag);
-      }
-
+      checkArgument(outputTags.contains(tag), "Unknown output tag %s", tag);
```

Comment:
I would assert that this is a bugfix, but it will mean that some pipelines crash that formerly did not.

@dhalperi @davorbonaci